### PR TITLE
Database Backend: Use configured serializer, instead of always pickling.

### PR DESCRIPTION
**Pull request by m4ddav3**
Use the BLOB as an sa.BLOB
Serialise the result an add to the db as bytes

### CODE REVIEW

**Comment by ask** [comment]
Is this backward compatible?

File: celery/backends/database/models.py [hunk 1]
```diff
@@ -57,7 +57,7 @@ class TaskSet(ResultModelBase):
     id = sa.Column(sa.Integer, sa.Sequence('taskset_id_sequence'),
                    autoincrement=True, primary_key=True)
     taskset_id = sa.Column(sa.String(255), unique=True)
-    result = sa.Column(PickleType, nullable=True)
```
Comment (ask) [inline celery/backends/database/models.py]: Should we then also remove PickleType, if it's no longer used?

Comment (m4ddav3) [inline celery/backends/database/models.py]: Sure, I've removed the line

**Comment by m4ddav3** [comment]
I need to check this, but I believe it is backward compatible while the result encoding type is set to 'pickle'. If the user has result encoding set to 'json' for instance, then it will change the behaviour (I suddenly found readable data that I can access easily)
I guess a script to convert the column to the required encoding could be provided.

**Comment by ask** [comment]
I was trying to merge this into master, and it does not pass tests,  it seems to be missing some critical parts, like deserialization of the task/group payloads :(

**Comment by m4ddav3** [comment]
Ugh. I need to dig up a Vagrant/Docker celery + rabbitmq + db environment

**Comment by ask** [comment]
@m4ddav3: `cd funtests/stress/run; vagrant up; vagrant shell`

**Comment by ask** [comment]
No database in that environment though, but shouldn't be too difficult to add

**Comment by m4ddav3** [comment]
@ask ,
I've had a dig around, and added tests for json serialisation with an sqlite db backend.
The existing tests expect that the serializer be able to handle stored Python objects, which I think isn't necessarily the right thing to do.

> Celery is written in Python, but the protocol can be implemented in any language

In order to keep in-line with the current methodology, but still be able to implement a JSON serializer, I suggest adding [jsonpickle](https://github.com/jsonpickle/jsonpickle). This should enable enough compatibility to the current scheme of encoding/decoding exceptions and/or other serializable Python objects. What do you think?

**Comment by ask** [comment]
jsonpickle would be nice, but sadly it has the same security concern as pickle, which is the ability to deserialize arbitrary objects by attribute path.

The tests must be rewritten to use json-safe data only, not sure how much work that is :(

**Comment by martincapello** [comment]
I created #5907, which is a rebased and slightly updated version of this PR.

**Pull request closed without merging**